src/engine/structures.rs

```diff
@@ -102,117 +166,112 @@ impl<T: HasName + HasUuid> Table<T> {
     }
 
     /// Get mutable item by name.
     pub fn get_mut_by_name(&mut self, name: &str) -> Option<&mut T> {
-        if let Some(index) = self.name_map.get(name) {
-            Some(&mut self.items[*index])
-        } else {
-            None
-        }
+        let uuid = match self.name_to_uuid.get(name) {
+            Some(uuid) => *uuid,
+            None => return None,
+        };
+        self.items
+            .get_mut(&uuid)
+            .map(|&mut (_, ref mut item)| item)
     }
 
     /// Get mutable item by uuid.
     pub fn get_mut_by_uuid(&mut self, uuid: Uuid) -> Option<&mut T> {
-        if let Some(index) = self.uuid_map.get(&uuid) {
-            Some(&mut self.items[*index])
-        } else {
-            None
-        }
+        self.items
+            .get_mut(&uuid)
+            .map(|&mut (_, ref mut item)| item)
     }
 
     /// Removes the item corresponding to name if there is one.
     pub fn remove_by_name(&mut self, name: &str) -> Option<T> {
-        if let Some(index) = self.name_map.remove(name) {
-            // Insert mappings for the about-to-be swapped element
-            {
-                let last_item = &self.items
-                                     .last()
-                                     .expect("name_map is non-empty <-> items is non-empty");
-                self.name_map.insert(last_item.name().into(), index);
-                self.uuid_map.insert(last_item.uuid(), index);
-            }
-
-            // Remove the item we want to remove and also the uuid mapping
-            let item = self.items.swap_remove(index);
-            self.uuid_map.remove(&item.uuid());
-
-            // Remove the name again, in case there is only one item.
-            self.name_map.remove(name);
-
-            Some(item)
+        if let Some(uuid) = self.name_to_uuid.remove(name) {
+            self.items.remove(&uuid).map(|(_, item)| item)
         } else {
             None
         }
     }
 
     /// Removes the item corresponding to the uuid if there is one.
     pub fn remove_by_uuid(&mut self, uuid: Uuid) -> Option<T> {
-        if let Some(index) = self.uuid_map.remove(&uuid) {
-            // Insert mappings for the about-to-be swapped element
-            {
-                let last_item = &self.items
-                                     .last()
-                                     .expect("uuid_map is non-empty <-> items is non-empty");
-                self.name_map.insert(last_item.name().into(), index);
-                self.uuid_map.insert(last_item.uuid(), index);
-            }
-
-            // Remove the item we want to remove and also the uuid mapping
-            let item = self.items.swap_remove(index);
-            self.name_map.remove(&*item.name());
-
-            // Remove the uuid again, in case there is only one item.
-            self.uuid_map.remove(&uuid);
-
+        if let Some((_, item)) = self.items.remove(&uuid) {
+            let name = self.name_to_uuid
```

Comment:
Maybe I'm missing something here, but when you remove the item you are getting the name and item right?  As you already know the name can't you just skip the `.ter().find ...` and go straight at `self.name_to_uuid.remove(&*name)`

Reply:
You're right! Nice catch!